---
title: "BSH Analysis"
author: "Sam McMillan"
date: "2023-11-30"
output: html_document
---

```{r setup, require=T, include=T}
library(tidyverse)
library(pheatmap)
library(ggpubr)
library(muscle)
library(curl)
library(rstatix)
library(janitor)
library(randomForest)
library(writexl)

sessionInfo()
```

# Get Dataframes
```{r}
samplekey<-readRDS("Dataframes/samplekey.rds")
genes_raw<-readRDS("Dataframes/genes_raw.rds")
bch_siglist<-readRDS("Dataframes/Stats/bch_siglist.rds")
bch_metadata<-readRDS("Dataframes/bch_metadata.rds")
bch_for_cor<-readRDS("Dataframes/bch_for_cor.rds")
MCBA_for_cor<-readRDS("Dataframes/MCBA_for_cor.rds")
MCBA_siglist<-readRDS("Dataframes/Stats/MCBA_siglist.rds")
tax_metadata<-readRDS("Dataframes/tax_metadata.rds")
ba_order<-readRDS("Dataframes/Bile acid order.rds")
mcba_order<-readRDS("Dataframes/MCBA order.rds")
tax_raw<-readRDS("Dataframes/tax_raw.rds")

source("Scripts/Color codes.R")
```

# Pull out BSHs
```{r}
BSH<-genes_raw%>%
  filter(str_detect(genefamily,"Choloylglycine|choloylglycine|Bile salt hydrolase"))%>%
  separate(genefamily,into=c("UniRef","taxa"),sep = "\\|",fill="right")%>%
  filter(!is.na(taxa))%>%
  separate(taxa,into=c("genus","species"),sep = "\\.",fill="right")%>%
  mutate(genus=str_remove(genus,"g\\_\\_"))%>%
  left_join(.,tax_metadata%>%select(phylum,family,genus)%>%unique(),by="genus")%>%
  select("UniRef","phylum","family","species",everything())%>%
  mutate(phylum=case_when(!is.na(phylum)~phylum, #Humann 3 does not have perfect matching to Metaphlan 4, These are ones that didn't overlap
                          genus %in% c("Tyzzerella","Lachnospiraceae_unclassified","Coprococcus","Lachnoclostridium","Subdoligranulum","Ruminococcus","Firmicutes_unclassified","Clostridiales_unclassified")~"Firmicutes", 
                          genus=="Methanobrevibacter" ~ "Euryarchaeota"))%>%
  mutate(family=case_when(!is.na(family)~family, 
                          genus %in% c("Tyzzerella","Lachnospiraceae_unclassified","Coprococcus","Lachnoclostridium")~"Lachnospiraceae",
                          genus=="Methanobrevibacter" ~ "Methanobacteriaceae",
                          genus %in% c("Subdoligranulum","Ruminococcus")~"Oscillospiraceae"))

##I want to get rid of mapped species that are not in the metaphlan output. There is currently disagreement between Metaphlan 4 SGBs and Humann 3 pangenomes. For example, C. butyricum genes are sometimes mapping to C. difficile genomes as Humann 3 uses pangenomes that include C. butyricum and C. difficile

tax_long<-tax_raw%>%
  filter(str_detect(clade_name,";s__"))%>%
  mutate(clade_name=str_remove(clade_name,".*s__"))%>%
  gather(key="CLIENT.IDENTIFIER",value="relabun",-clade_name)%>%
  mutate(clade_name=str_replace_all(clade_name,"_"," "))


BSH_long<-BSH%>%
  gather(key="CLIENT.IDENTIFIER",value="CPM",-c("UniRef","phylum","family","genus","species"))%>%
  mutate(CLIENT.IDENTIFIER=factor(CLIENT.IDENTIFIER,levels=levels(samplekey$CLIENT.IDENTIFIER)))%>%
  left_join(.,samplekey,by="CLIENT.IDENTIFIER")%>%
  mutate(species=str_remove(species,"s__"))%>%
  mutate(species=str_replace_all(species,"_"," "))

BSH_filt<-BSH_long%>%
  left_join(.,tax_long,by=c("CLIENT.IDENTIFIER","species"="clade_name"))%>%
  filter(relabun>0)%>%
  mutate(CLIENT.IDENTIFIER=factor(CLIENT.IDENTIFIER,levels=levels(samplekey$CLIENT.IDENTIFIER)))%>%
  select(-SUBJ.ID,-TIME.POINT.2,-prepost,-relabun)%>%
  complete(CLIENT.IDENTIFIER,fill=list(UniRef="drop",phylum="drop",family="drop",species="drop",genus="drop",CPM=0))%>%
  spread(key=CLIENT.IDENTIFIER,value=CPM,fill=0)%>%
  filter(UniRef!="drop")


#for writing
length(unique(BSH$species))-1 #-1 for the blasted only BSHs

BSH_long%>%filter(CPM==0)%>%group_by(CLIENT.IDENTIFIER)%>%summarize(count=n())%>%arrange(-count)

BSH_long%>%select(SUBJ.ID,TIME.POINT.2,CPM)%>%group_by(SUBJ.ID,TIME.POINT.2)%>%summarize(sum=sum(CPM))%>%spread(key=TIME.POINT.2,value=sum)

saveRDS(BSH,"Dataframes/BSH.rds")
```


# Boxplot
```{r}
BSH_by_time<-BSH_long%>%
  select(CLIENT.IDENTIFIER,CPM,TIME.POINT.2,SUBJ.ID)%>%
  group_by(CLIENT.IDENTIFIER,TIME.POINT.2,SUBJ.ID)%>%
  summarize(sum=sum(CPM))

BSH_boxplot_stats<-BSH_by_time%>%
  ungroup()%>%
  select(sum,TIME.POINT.2,SUBJ.ID)%>%
  complete(TIME.POINT.2, SUBJ.ID, fill=list(sum=NA))%>%
  arrange(SUBJ.ID)%>%
  pairwise_wilcox_test(sum ~ TIME.POINT.2,
                       p.adjust.method="holm",
                       ref.group = "Pre",
                       paired=T)%>%
  add_xy_position(x="TIME.POINT.2")

ggplot(BSH_by_time,aes(x=TIME.POINT.2,y=sum,group=TIME.POINT.2))+
  geom_dotplot(aes(fill=TIME.POINT.2),binaxis="y",stackdir="center",binwidth=325/50,dotsize=1,stroke=0.5)+
  geom_boxplot(alpha=0,linewidth=0.3)+
  theme_bw()+
  theme(text=element_text(size=8),
        legend.position="none",
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x= element_blank())+
  labs(y = "Total BSH Abundance (CPM)",
       x= "",
       fill="Time Post FMT")+
  scale_y_continuous(limits=c(0,325),expand=c(0,0))+
  scale_fill_manual(values=anno_color$TIME.POINT.2)+
  stat_pvalue_manual(filter(BSH_boxplot_stats,p.adj<0.05), 
                     label = "p.adj.signif",
                     tip.length=0.01,
                     step.increase=0.05,
                     inherit.aes=F)

ggsave("Figures/BSH/BSH_boxplot.pdf",height=2.1,width=2.25, units="in")

saveRDS(BSH_boxplot_stats,"Dataframes/Stats/All BSH boxplot wilcox signed rank holm.rds")
```

# Get Sequences for BSHs, Align, ID loop

```{r}
### What's going on with the loop?
uni<-matrix(ncol=8,nrow=0)
colnames(uni)<-c("Entry","Sequence","Catalytic.activity","Pathway","EC.number","Gene.Ontology.IDs","GeneID","KEGG")

BSH_genes<-BSH%>%
  select(UniRef)%>%
  mutate(UniRef=str_remove_all(UniRef,"UniRef90_|\\: ?(.*)"))%>%
  unique()

#There's probably a better way to do this query
for(i in 1:nrow(BSH_genes)){
  uni<-rbind(uni,read.delim(curl(paste0("https://rest.uniprot.org/uniprotkb/search?query=%28accession%3A",BSH_genes[i,1],"%29&fields=accession,sequence,cc_catalytic_activity,cc_pathway,ec,go_id,xref_geneid,xref_kegg&format=tsv")),sep="\t"))
}
write.csv(uni,"Data and Metadata/bsh_genes_uni_raw.csv",row.names=F)

##I had to manually pull the ones in UniParc
uni<-read.csv("Data and Metadata/bsh_genes_uni.csv")

aaseq<-AAStringSet((uni%>%column_to_rownames("Entry"))$Sequence)
aligner<-muscle(aaseq)

alignment<-as.data.frame(aligner@unmasked)%>%
        mutate(loci=str_sub(x,480,482))%>%
        mutate(location=480-str_count(str_sub(x,0,479),"-"))%>%
        mutate(Sequence=str_remove_all(x,"-"))%>%
        mutate(loci2=str_sub(Sequence,location-1,location+2))%>%
        left_join(uni,.,by="Sequence")%>%
        mutate(motif=if_else(str_detect(loci,"G.G"),"Taurine-preferring","No Loop"))%>%
        mutate(motif=if_else(str_detect(loci,"SR."),"Glycine-preferring",motif))
```

# Plot Loops by time
```{r}
BSH_loops<-BSH%>%
  mutate(UniRef=str_remove_all(UniRef,"UniRef90_|\\:.*"))%>%
  left_join(.,alignment,by=c("UniRef"="Entry"))%>%
  select(motif,genus,family,species,samplekey$CLIENT.IDENTIFIER)%>%
  gather(key="CLIENT.IDENTIFIER",value="CPM",-motif,-genus,-species,-family)%>%
  left_join(.,samplekey,by="CLIENT.IDENTIFIER")%>%
  mutate(motif=factor(motif,levels=c("Glycine-preferring","Taurine-preferring","No Loop")))%>%
    select(motif,CLIENT.IDENTIFIER,CPM,TIME.POINT.2,SUBJ.ID)%>%
    group_by(motif,CLIENT.IDENTIFIER,TIME.POINT.2,SUBJ.ID)%>%
  summarize(sum=sum(CPM))
  
BSH_loops_stats<-BSH_loops%>%
  ungroup()%>%
  select(sum,TIME.POINT.2,SUBJ.ID, motif)%>%
  complete(TIME.POINT.2, SUBJ.ID,motif, fill=list(sum=NA))%>%
  arrange(SUBJ.ID)%>%
  group_by(motif)%>%
  pairwise_wilcox_test(sum ~ TIME.POINT.2,
                       p.adjust.method="holm",
                       ref.group = "Pre",
                       paired=T)%>%
  add_xy_position(x="TIME.POINT.2")%>%
  mutate(y.position=y.position-60)

ggplot(data=BSH_loops,aes(x=TIME.POINT.2,y=sum,fill=TIME.POINT.2))+
  geom_dotplot(binaxis="y",stackdir="center",binwidth=260/50,dotsize=1,stroke=0.5)+
  geom_boxplot(alpha=0,linewidth=0.3)+
  facet_wrap(~motif)+
  theme_bw()+
  theme(text=element_text(size=8),
        legend.position="none",
        axis.text.x=element_text(angle=-90),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x= element_blank(),
        strip.background=element_blank())+
  labs(y = "Total BSH Abundance (CPM)",
       x= "",
       fill="Time Post FMT")+
  scale_y_continuous(expand=c(0,2,0,5),breaks=c(0,100,200,300,400))+
  scale_fill_manual(values=anno_color$TIME.POINT.2)+
  stat_pvalue_manual(filter(BSH_loops_stats,p.adj<0.05), 
                     label = "p.adj.signif",
                     tip.length=0.01,
                     step.increase=0.05,
                     step.group.by="motif",
                     inherit.aes=F)


ggsave("Figures/BSH/BSH loop mean plot.pdf",height=2.5,width=2.7,units="in") 

saveRDS(BSH_loops_stats,"Dataframes/Stats/BSH loop boxplot wilcox signed rank holm.rds")
```


# Family Plot BSHs
```{r}
#Make sure we have colors for everyone
unique(BSH$family)[!unique(BSH$family) %in% names(anno_color$familyplot)]

BSH_family<-BSH%>%
  gather(CLIENT.IDENTIFIER,CPM,-c(UniRef,phylum,species,genus,family))%>%
  mutate(CLIENT.IDENTIFIER=factor(CLIENT.IDENTIFIER, levels=levels(samplekey$CLIENT.IDENTIFIER)))%>%
  mutate(family=str_remove(family,"f__"))%>%
  mutate(family=str_replace_na(family,"Other"))%>%
  mutate(family=factor(family,levels=names(anno_color$familyplot)))%>%
  droplevels()

ggplot(BSH_family,aes(x=CLIENT.IDENTIFIER,y=CPM, fill=family))+
  geom_bar(stat="identity")+
  theme_bw()+
  theme(axis.text.x=element_blank(),
        text=element_text(size=8),
        legend.position="right",
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x= element_blank(),
        legend.key.size = unit(0.1,"inches"))+
  scale_fill_manual(values = anno_color$familyplot)+
  scale_y_continuous(limits=c(0,250),expand=c(0,0))+
  labs(x="",y="BSH Abundance (CPM)",fill="Family")
ggsave("Figures/BSH/BSH_family_plot.pdf",height=2,width=3.2, units="in")
```

#Species Heatmap
```{r}
##All species, supplemental figure
BSH_hmap_all<-BSH%>%
        select(species,samplekey$CLIENT.IDENTIFIER)%>%
        gather(CLIENT.IDENTIFIER,CPM,-species)%>%
        mutate(CLIENT.IDENTIFIER=factor(CLIENT.IDENTIFIER,levels=samplekey$CLIENT.IDENTIFIER))%>%
        filter(CPM!=0)%>%
        group_by(species,CLIENT.IDENTIFIER)%>%
        summarise(CPM=sum(CPM))%>%
        mutate(CPM=log10(CPM+1))%>% #Transform for plot
        complete(CLIENT.IDENTIFIER,fill=list(Species="Other",CPM=0))%>% #Some samples didn't have BSHs
        spread(CLIENT.IDENTIFIER,CPM)%>%
        select(species,samplekey$CLIENT.IDENTIFIER)%>%
        left_join(.,BSH%>%select(species,family)%>%unique(),by="species")%>%
        mutate(family=replace_na(family,"Other"))%>%
        mutate(species=str_remove(species,"s__"))%>%
        mutate(species=str_replace_all(species,"_"," "))%>%
        mutate(species=replace_na(species,"Other"))%>%
        column_to_rownames("species")

pheatmap(BSH_hmap_all[1:45],
         annotation_col=samplekey%>%column_to_rownames("CLIENT.IDENTIFIER")%>%select(TIME.POINT.2),
         annotation_row=BSH_hmap_all%>%select(family),
         annotation_colors = list(TIME.POINT.2=anno_color$TIME.POINT.2, family=anno_color$familyplot[names(anno_color$familyplot) %in% BSH_hmap_all$family]),
         cluster_cols=F,
         cluster_rows=T,
         show_colnames = F,
         annotation_names_col = F,
         annotation_names_row = F,
         fontsize=12,
         fontsize_row=12,
         color=colorRampPalette(c("white","grey","black"),bias=2)(50),
         filename="Figures/Supplementals/BSH_all_heatmap.pdf",
         border_color="grey20",
         drop_levels=T,
         width=11,
         height=12
)

```

# Correlation Repeated measures
```{r}
BSH_for_cor<-BSH%>%
        select(species,UniRef,samplekey$CLIENT.IDENTIFIER)%>%
        mutate(species=str_remove(species,"s__"))%>%
        mutate(species=str_replace_all(species,"_"," "))%>%
        mutate(species=replace_na(species,"Other"))%>%
        mutate(UniRef=str_remove_all(UniRef,"UniRef90_|\\: .*"))%>%
        mutate(BSH=paste(UniRef, species),.keep="unused")%>%
        column_to_rownames("BSH")%>%
        t()
      
saveRDS(BSH_for_cor,"Dataframes/BSH_for_cor.rds")

source("Scripts/pheatmap_left.R")

source("Scripts/rm_corr_all.R")

rmcor_MCBA_BSH<-rmcorr_test(samplekey%>%column_to_rownames("CLIENT.IDENTIFIER"),
                           "SUBJ.ID",
                           BSH_for_cor,
                          MCBA_for_cor[mcba_order],
                          "BH")


MCBA_BSH_cor_plot<-rmcor_MCBA_BSH%>%
  dplyr::rename("BA"=df2,"BSH"=df1)%>%
  mutate(BA=factor(BA, levels=rev(mcba_order)))%>%
  filter(!is.na(BA))%>%
  filter(BSH %in% (filter(.,p.adj<=0.05))$BSH)%>%
  select(-p,-p.adj)%>%
  group_by(BSH)%>%
  spread(key=BSH,value=r,fill=0)%>%
  arrange(BA)%>%
  column_to_rownames("BA")

MCBA_BSH_cor_p<-rmcor_MCBA_BSH%>%
  dplyr::rename("BA"=df2,"BSH"=df1)%>%
  mutate(BA=factor(BA, levels=rev(mcba_order)))%>%
  filter(!is.na(BA))%>%
  #filter(p.adj<=0.05)%>%
  filter(BSH %in% (filter(.,p.adj<=0.05))$BSH)%>%
  mutate(ast=symnum(p.adj,cutpoints=c(0,0.05,1),symbols=c("*","")))%>%
  select(-r,-p,-p.adj)%>%
  group_by(BSH)%>%
  spread(key=BSH,value=ast,fill="")%>%
  arrange(BA)%>%
  column_to_rownames("BA")

MCBA_BSH_anno<-as.data.frame(colnames(MCBA_BSH_cor_plot))%>%
  dplyr::rename("BSH"=1)%>%
  separate(BSH,into=c("UniRef","species"),sep=" ",extra="merge",remove=F)%>%
  left_join(.,BSH%>%
              select(species,family)%>%
              mutate(species=str_remove(species,"s__"))%>%
              mutate(species=str_replace_all(species,"_"," "))%>%
              unique(),by="species")%>%
  column_to_rownames("BSH")%>%
  mutate(family=factor(family,levels=names(anno_color$familyplot)))

fam_list<-names(anno_color$familyplot)[names(anno_color$familyplot) %in% unique(MCBA_BSH_anno$family)]
for (f in 1:length(fam_list)){
    d<-MCBA_BSH_cor_plot[which(colnames(MCBA_BSH_cor_plot) %in% rownames(filter(MCBA_BSH_anno,family==fam_list[f])))]
    p<-MCBA_BSH_cor_p[which(colnames(MCBA_BSH_cor_plot) %in% rownames(filter(MCBA_BSH_anno,family==fam_list[f])))]
    rn=F
    w=1/6*ncol(d)+0.1
    l=F
  
  if(ncol(d)==1){clust=F}else{clust=T}
    
  if(f==1){
    rn=T 
    w=w+1.2}

  if(f==14){
    l=T
    w=w+0.6}
    
  inames<-  as_tibble(colnames(d))%>%
    separate(value, into=c("UniRef","strain"),sep=" ",extra="merge",remove=F)
  
  names<-mapply(function (x,y) bquote(plain(.(x))~italic(.(y))),inames$UniRef,inames$strain)
  
pheatmap(d,
         annotation_col=MCBA_BSH_anno%>%select(family),
         display_numbers=p,
         labels_col=as.expression(names),
         number_color="white",
         cluster_distance_cols="correlation",
         cluster_rows=F,
         cluster_cols=clust,
         annotation_colors=list(family=anno_color$familyplot[which(names(anno_color$familyplot) %in% unique(MCBA_BSH_anno$family))]),
         annotation_names_col=F,
         show_rownames=rn,
         fontsize=12,
         cellheight = 12,
         cellwidth = 12,
         legend=l,
         annotation_legend=F,
         breaks=seq(-1,1,2/60),
 color=c(rev(colorRampPalette(c("grey88","dodgerblue","dodgerblue3"),bias=1)(30)),colorRampPalette(c("grey88","red", "red3"),bias=1)(30)),
         filename=paste0("Figures/BSH/MCBA_vs_BSH_rmcor_BH_RF_",f,".pdf"),
         width=w,
         height=12,
         angle_col=270)
}

##Reset the pheatmap left overwrite
detach("package:pheatmap", unload = TRUE)
library(pheatmap)

saveRDS(rmcor_MCBA_BSH,"Dataframes/Stats/BSH vs MCBA rmcor.rds")

#for writing
colnames(MCBA_BSH_cor_plot)%>%
  as_tibble()%>%
  separate(value, into="UniProt",extra="drop")%>%
  group_by(UniProt)%>%
  summarize(count=n())%>%
  arrange(-count)%>%
  head(10)
```

#Data export for Mendeley
```{r data export}
write_xlsx(list("6A"=BSH_by_time,
                "6A Wilcoxon signed rank"=BSH_boxplot_stats,
                "6B"=BSH_loops,
                "6B Wilcoxon signed rank"=BSH_loops_stats,
                "6C"=BSH_family,
                "6D"=MCBA_BSH_cor_plot,
                "6D linear model"=rmcor_MCBA_BSH),
           "Excel Exports/Figure 6.xlsx")

write_xlsx(list("S6"=BSH_hmap_all),
           "Excel Exports/Figure S6.xlsx")

```


